sys-apps/kmod: switch to meson, drop static-lib support

evelikov opened the pull request:
Over the last few months, we've got meson support upstream and the plan is to quickly migrate from autotools with just one version overlap.

There is no plan to support static-lib upstream and as illustrated in the original thread it doesn't quite work since we get lots of symbol collisions, so drop the IUSE flag and local workarounds.

Upstream has also started pulling the shell completions into the main project, to indicate that use the respective get_*compdir helpers.

Suspecting those could live behind an IUSE flag, as does scdoc... Which can come at a later day.

<!-- Please put the pull request description above -->

---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [ ] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

---

NOTE: This PR is done as a heads-up and is completely untested and likely broken. Please take as much or as little of it as you see fit.

Couple of questions, if I may:
 - Reading through the wiki [1], it's not entirely clear if `split-usr` is planed to be phased out. Any references?
 - Gentoo (and Debian fwiw) have been carrying `softdep Xhci_hcd pre: ehci_hcd` quirk for decade, while we have a _similar_ change in the upstream kernel since 2013 [2]. Is the local workaround still needed, any references?

Thanks in advance o/

[1] https://wiki.gentoo.org/wiki/Merge-usr
[2] https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=05c92da0c52494caf97d58be1b05b1f95a79ce4e

thesamesam commented:
Thank you again for your care in helping distros with the upstream work, it's really appreciated.

eli-schwartz reviewed (commented):
> Reading through the wiki [1], it's not entirely clear if `split-usr` is planed to be phased out. Any references?

There is no compelling desire to change anything here, in large part because most of the people who find it annoying are happy enough not thinking about it as long as systemd is used. Since https://www.gentoo.org/support/news-items/2024-01-05-usr-initramfs.html it is no longer supported in the general case to have libraries in /lib on split-usr systems due to the support complexity tarpit -- split-usr itself still "works" in the sense that people seem attached to the idea and Gentoo continues to install programs to their canonical location even if that is /bin.

The kernel will somehow have to call programs at the right location on such systems... but meson does support a bindir / sbindir that is outside of `--prefix` so I think this shouldn't be a big deal.

thesamesam commented:
See eb13455af714ffb37726b04cbc87f91bbc2fbf97 and e0f3f9da51b19053e04e59e723257f357bddb5d2.

floppym commented on sys-apps/kmod/kmod-9999.ebuild:
```diff
@@ -41,73 +41,33 @@ RDEPEND="
+		-Dfishcompletiondir="$(get_fishcompdir)"
+		-Dzshcompletiondir="$(get_zshcompdir)"
+		$(meson_use debug debug-messages)
+		$(meson_use docs manpages)
```
This seems wrong for 2 reasons:

1. The use flag is named "doc", not "docs".
2. It should not control the manpages.

evelikov replied:
Silly copy/paste mistake - thanks it's fixed now.

floppym commented on sys-apps/kmod/kmod-9999.ebuild:
```diff
@@ -41,73 +41,33 @@ RDEPEND="
-		$(use_with zlib)
-		$(use_with zstd)
+	local emesonargs=(
+		--prefix "${EPREFIX}"
```
This is wrong. Should be "${EPREFIX}/usr", which is the default supplied by the eclass.

We should override bindir and sbindir instead.

evelikov replied:
Also fixed, thanks.

It seem to work properly with split-usr atm... We either have a bug or there one in meson itself. Will investigate and report back.

evelikov replied:
With the following PR things should just work (tm). That said, I haven't tested it so use your judgement before blindly testing it :-) 

https://github.com/kmod-project/kmod/pull/205

floppym commented on sys-apps/kmod/kmod-9999.ebuild:
```diff
@@ -117,15 +77,6 @@ src_install() {
 		done
 	fi
 
-	cat <<-EOF > "${T}"/usb-load-ehci-first.conf
-	softdep uhci_hcd pre: ehci_hcd
-	softdep ohci_hcd pre: ehci_hcd
-	EOF
-
-	insinto /lib/modprobe.d
-	# bug #260139
-	doins "${T}"/usb-load-ehci-first.conf
-
```
Why are you dropping this?

floppym replied:
Never mind, I see this is explained in the commit message.

floppym reviewed (changes requested):
As was stated in the original comment, this is clearly untested and broken. I am adding review comments in case someone actually uses it as a base.

evelikov commented:
Rebased and updated to handle the latest upstream - symlinks are now installed in sbin.

Fwiw I don't plan on making any more distribution visible changes for kmod v34. If anyone wants to check this PR and land anything with functionally similar that would be awesome.

floppym reviewed (approved):
It looks ok to me. Just needs testing before we merge.
